__tests__/integration/repeatedPropertyURI.js

```diff
@@ -0,0 +1,17 @@
+// Copyright 2019 Stanford University see LICENSE for license
+
+import pupExpect from 'expect-puppeteer'
+import { testUserLogin } from './loginHelper'
+
+describe('A repeated propertyURI', () => {
+  beforeAll(async () => {
+    await testUserLogin()
```

Comment:
fwiw, i found that tests consistently passed for me without a `return` here.  which is what i'd sorta expected, since it doesn't seem like anything cares about the value to which the anonymous `beforeAll` promise resolves to (said promise being returned by the anonymous `async` function passed into it on line 7).  the `await` on the `testUserLogin` promise seemed like the crux of it, since that's what gates the resolving of the promise passed to the `beforeAll` (that promise resolves when the anon function finishes executing).

if people really want, i can add the `return` in, but i just wanted to note that the deviation from elsewhere is intentional.

if the above logic is sound, i'd also gently urge us to try getting rid of similar unnecessary `return`s, since they always seemed slightly misleading to me (but it's for sure not something i'd prioritize, since it's functionally harmless and it works).

Reply:
as I understand it, the "return" ensures the promise is resolved before returning, as opposed to just firing off an async thing and NOT paying attention.

https://jestjs.io/docs/en/setup-teardown

```
beforeEach and afterEach can handle asynchronous code in the same ways that tests can handle asynchronous code - they can either take a done parameter or return a promise. For example, if initializeCityDatabase() returned a promise that resolved when the database was initialized, we would want to return that promise:

beforeEach(() => {
  return initializeCityDatabase();
});
```

So I hear you about "it's an async anyway" but I don't know the jest code well enough to know if it properly handles the async, as beforeEach() is not inherently async.  And if we *need* the return, then let's use it consistently.  I'd wonder if multiple await statements in a beforeEach would *all* be executed without the return.  I'd hope so, in which case I'd happily get rid of the "return" in the beforeXXX.

Reply:
> as I understand it, the "return" ensures the promise is resolved before returning, as opposed to just firing off an async thing and NOT paying attention.

my understanding of this differs.  afaict, the only thing the `return` does is specify the value to which the implicit wrapping `Promise` resolves if the `async` function completes successfully.  the `return` doesn't gate any action, or force any waiting for anything, unless that waiting is conditional on the value to which the `Promise` resolves.  but i _think_ all we care about is that the `Promise` given to `beforeEach`, `beforeAll`, `it`, etc resolves if and only if the setup or test code has completed.  by definition, the `async` function `Promise` will only resolve when the body of the function completes:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#Return_value
> ### Return Value
> A `Promise` which will be resolved with the value returned by the async function, or rejected with an uncaught exception thrown from within the async function.

> don't know the jest code well enough to know if it properly handles the async

i think from the perspective of the Jest functions, it's just getting a `Promise`.  it doesn't care whether that's because it was given an async function that always implicitly returns a `Promise`, or if it's because a "regular" function explicitly returned a `Promise` object (of its creation, from a method call, etc).

>  And if we need the return, then let's use it consistently

i would agree, but i guess i have the opposite conclusion, that it's not necessary, and that we should not use it "just in case", because that's subtly misleading.  i would be in favor of removing it where it's not needed, but i also wouldn't consider that a high priority, because other than being superfluous, it's harmless.  i don't think anything cares one way or the other about the values to which these promises resolve.

>  I'd wonder if multiple await statements in a beforeEach would all be executed without the return.

i believe they will, and i wrote a bit of test code last night to illustrate this:  https://gist.github.com/jmartin-sul/6f23e9d61b6a6f986efa99306b73ef14

happy to pair on running it (if only to be shown that i'm wrong if i am).  but ultimately, i _think_ the `waitWaitWait` part of that gist shows that there is no need for a `return` value to have `await` block in an `async` function, and i also think it shows that having an explicit `return` does not block on the completion of the `async` function unless that function call is explicitly `await`ed (e.g. `returnsAwaitResult()` and `noReturnValue()` both immediately return a `Promise` upon execution, even though neither function body will finish for a number of seconds after the function execution starts.  you can see from playing with that code that the returned `Promise` will be resolved as soon as the function body completes, but not before.

see also:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

> The `await` expression causes `async` function execution to pause until a `Promise` is settled, that is fulfilled or rejected, and to resume execution of the `async` function after fulfillment. When resumed, the value of the `await` expression is that of the fulfilled `Promise`.

which is to say that returning the value of the `await` does not return a `Promise` (because the result of an `await` is not a `Promise`, but the value to which the `await`ed `Promise` resolves).  an `async` method can't ever return anything other than a `Promise`

you can see some illustration of the return types of `await` statements and `async` functions in this other way-too-verbose comment:  https://github.com/LD4P/sinopia_profile_editor/pull/213#issuecomment-493646501

but i think the gist in this comment illustrates the `await` blocking better and more concisely.

as a side-note, i don't think it's particularly useful to think of a function as `async` if it is not declared as such.  `async` is just syntactic sugar for wrapping a function in a `Promise` and letting you wait for other `Promise`s that are kicked off in said function body, without so much callback boilerplate.  a function is `async` if and only if it's declared as such, regardless of whether it deals with `Promise`s internally or does asynchronous stuff inside of itself.

Reply:
Thank you for the detailed explanation; I believe I am finally starting to understand, rather than trying to parrot the exact language in jest docs without fully understanding.   I *would* like to pair or o.w. go through the examples empirically w you as I think it will help solidify my understanding of how this promise stuff and async stuff works.

Feel free to make a ticket about converting existing async tests to do the right thing in their setup/teardown.

Reply:
will do, thanks for reading and sanity checking!

Reply:
filed #840